src/microengine_utils/scanner.py

```diff
@@ -0,0 +1,148 @@
+from .errors import BaseScanError, CalledProcessScanError
+
+
+async def create_scanner_exec(
+    *cmd: 'str',
+    stdout=asyncio.subprocess.PIPE,
+    stderr=asyncio.subprocess.PIPE,
+    stdin=asyncio.subprocess.DEVNULL,
+    check: 'bool' = False,
+):
+    """Run an engine filescan `cmd`, timing out after `timeout` seconds"""
+    try:
+        proc = await asyncio.subprocess.create_subprocess_exec(
+            *cmd,
+            stdout=stdout,
+            stderr=stderr,
+            stdin=stdin,
+        )
+        streams = await proc.communicate()
+        if check and proc.returncode != 0:
+            raise CalledProcessScanError(cmd, f'Non-zero return code: {proc.returncode}')
+        sout, serr = (s.decode(errors='ignore') for s in streams)
+        return proc.returncode, sout, serr
+    except (FileNotFoundError, BrokenPipeError, ConnectionResetError) as e:  # noqa
+        proc.kill()
+        raise CalledProcessScanError(cmd, str(type(e)))
+
+
+def scanalytics(statsd: 'datadog.DogStatsd' = datadog.statsd, engine_info: 'Optional[EngineInfo]' = None):
+    """Decorator for `async_scan` to automatically handle errors and boilerplate scanner metadata
+
+    - Record and send timing data to Datadog
+    - Read the `ScanResult`'s fields to automatically figure out which metrics should be collected
+    - Merges `ScanResult` `metadata` with boilerplate scanner information from `EngineInfo`
+    """
+    verbose: 'bool' = bool(os.getenv('MICROENGINE_VERBOSE_METRICS', False))
+
+    def wrapper(scan_fn: 'Callable') -> 'Callable':
+        def collect_success(scan: 'ScanResult', tags: 'List[str]') -> 'ScanResult':
+            # invalid bit or verdict
+            if type(scan.verdict) is not bool or type(scan.bit) is not bool:
+                statsd.increment(SCAN_TYPE_INVALID, tags=tags)
+
+            # successful scan, verdict reported
+            elif scan.bit:
+                with suppress(AttributeError, KeyError):
+                    getter = itemgetter if isinstance(scan.metadata, Mapping) else attrgetter
+                    tags.append(f'malware_family:{getter("metadata.malware_family")(scan)}')
+
+                # malicious/benign metrics
+                if verbose:
+                    statsd.increment(
+                        SCAN_VERDICT, tags=[*tags, 'verdict:malicious' if scan.verdict else 'verdict:benign']
+                    )
+
+                statsd.increment(SCAN_SUCCESS, tags=tags)
+
+            # no result reported
+            elif not scan.bit:
+                if verbose:
+                    statsd.increment(SCAN_NO_RESULT, tags=tags)
+
+            return scan
+
+        def collect_error(err: 'BaseScanError', tags: 'List[str]') -> 'ScanResult':
+            # If we encountered a scan error, we still return a scan result with `scan_error`
+            # documenting the error which occurred
+            v = Verdict().set_malware_family('').add_extra('scan_error', err.event_name)
```

Comment:
If bit is false this metadata won't get read anywhere

Reply:
This is still really useful, even if we don't end up using the metadata. I have already found it really helpful in debugging scan tests, which now report _why_ they are failing, rather than a naked `ScanResult`.